File: we_reader/lib/reader/pagination.dart
import 'package:flutter/material.dart';

import '../models/models.dart';

/// 一页的定位信息：属于哪一章、章节内字符区间。
class PageSpec {
  const PageSpec({
    required this.chapterIndex,
    required this.start,
    required this.end,
  });

  final int chapterIndex;

  /// 章节内起始字符偏移（含）。
  final int start;

  /// 章节内结束字符偏移（不含）。
  final int end;

  bool containsOffset(int offset) => offset >= start && offset < end;
}

/// 分页引擎：用 [TextPainter] 逐行排版，把每章切成整行对齐的页。
class Paginator {
  Paginator._();

  /// 把整本书按当前排版参数切页。示例书籍很短，全量分页开销可忽略；
  /// 长书应改为按章懒分页。
  static List<PageSpec> paginateBook({
    required List<Chapter> chapters,
    required TextStyle style,
    required Size pageSize,
  }) {
    final pages = <PageSpec>[];
    for (var i = 0; i < chapters.length; i++) {
      final ranges = paginateChapter(
        text: chapters[i].content,
        style: style,
        pageSize: pageSize,
      );
      for (final range in ranges) {
        pages.add(
          PageSpec(chapterIndex: i, start: range.start, end: range.end),
        );
      }
    }
    return pages;
  }

  /// 对单章文本分页，返回章节内字符区间列表。
  static List<TextRange> paginateChapter({
    required String text,
    required TextStyle style,
    required Size pageSize,
  }) {
    if (text.isEmpty) return [const TextRange(start: 0, end: 0)];

    final painter = TextPainter(
      text: TextSpan(text: text, style: style),
      textDirection: TextDirection.ltr,
    )..layout(maxWidth: pageSize.width);

    final lines = painter.computeLineMetrics();
    final ranges = <TextRange>[];
    double pageTop = 0;
    double lineTop = 0;
    int pageStart = 0;

    for (final line in lines) {
      final lineBottom = lineTop + line.height;
      // 当前行放不进本页时，在该行前断页。
      if (lineBottom - pageTop > pageSize.height + 0.5 && lineTop > pageTop) {
        final breakOffset = painter
            .getPositionForOffset(Offset(0, lineTop + line.height / 2))
            .offset;
        if (breakOffset > pageStart) {
          ranges.add(TextRange(start: pageStart, end: breakOffset));
          pageStart = breakOffset;
          pageTop = lineTop;
        }
      }
      lineTop = lineBottom;
    }
    if (pageStart < text.length) {
      ranges.add(TextRange(start: pageStart, end: text.length));
    }
    painter.dispose();
    return ranges.isEmpty ? [TextRange(start: 0, end: text.length)] : ranges;
  }
}

File: we_reader_rn/src/reader/pagination.ts
import type { TextStyle } from 'react-native';
import type { Chapter, PageSpec } from '../types/models';
import type { MeasuredLine } from './TextMeasurer';

export type Measure = (text: string, style: TextStyle, width: number) => Promise<MeasuredLine[]>;

interface FlatLine {
  start: number; // chapter-relative
  end: number;
  height: number;
}

async function measureChapterLines(
  measure: Measure,
  content: string,
  style: TextStyle,
  width: number
): Promise<FlatLine[]> {
  const paragraphs = content.length > 0 ? content.split('\n') : [];
  const flat: FlatLine[] = [];
  let paragraphOffset = 0;
  for (const paragraph of paragraphs) {
    const lines = paragraph.length > 0 ? await measure(paragraph, style, width) : [{ length: 0, height: style.lineHeight ?? 20 }];
    let cursor = paragraphOffset;
    for (const line of lines) {
      flat.push({ start: cursor, end: cursor + line.length, height: line.height });
      cursor += line.length;
    }
    paragraphOffset += paragraph.length + 1; // +1 for the joining '\n'
  }
  return flat;
}

/**
 * Breaks a chapter's lines into pages that fit `pageHeight`, cutting only
 * on whole-line boundaries (never mid-line) — mirrors the original app's
 * `Paginator.paginateChapter`.
 */
function buildPages(lines: FlatLine[], chapterIndex: number, contentLength: number, pageHeight: number): PageSpec[] {
  if (lines.length === 0) {
    return [{ chapterIndex, start: 0, end: contentLength }];
  }
  const pages: PageSpec[] = [];
  let pageStart = lines[0].start;
  let usedHeight = 0;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (usedHeight + line.height > pageHeight && usedHeight > 0) {
      pages.push({ chapterIndex, start: pageStart, end: line.start });
      pageStart = line.start;
      usedHeight = 0;
    }
    usedHeight += line.height;
  }
  pages.push({ chapterIndex, start: pageStart, end: contentLength });
  return pages;
}

export async function paginateChapter(
  measure: Measure,
  chapter: Chapter,
  chapterIndex: number,
  style: TextStyle,
  width: number,
  pageHeight: number
): Promise<PageSpec[]> {
  const lines = await measureChapterLines(measure, chapter.content, style, width);
  return buildPages(lines, chapterIndex, chapter.content.length, pageHeight);
}

export async function paginateBook(
  measure: Measure,
  chapters: Chapter[],
  style: TextStyle,
  width: number,
  pageHeight: number
): Promise<PageSpec[]> {
  const all: PageSpec[] = [];
  for (let i = 0; i < chapters.length; i++) {
    const pages = await paginateChapter(measure, chapters[i], i, style, width, pageHeight);
    all.push(...pages);
  }
  return all;
}
